### PR TITLE
[#43] [Backend] As a user, I can see the available surveys on the home screen

### LINKED ISSUE #43: [Backend] As a user, I can see the available surveys on the home screen
## Why

The user can see a list of the surveys which have not been finished by the users on the home screen. The user can only see one item on the list at a time.

## Acceptance criteria

Call the API `/api/v1/surveys` to get the list of surveys, use default `page[number]=1` and `page[size]=10`.

## Resources

- [API document for /api/v1/surveys](https://documenter.getpostman.com/view/11835486/T1LJjTqj#2dd12ea2-3ecb-46d5-a48b-64256346e4e1)

### PR DESCRIPTION
**Pull request by nmint8m**
- Close #43

## What happened 👀

The user can see a list of the surveys which have not been finished by the users on the home screen. The user can only see one item on the list at a time.

Call the API `/api/v1/surveys` to get the list of surveys, use default `page[number]=1` and `page[size]=10`.

## Insight 📝

`N/A`

## Proof Of Work 📹

image

image

### CODE REVIEW

File: lib/api/api_service.dart [hunk 1]
```diff
@@ -14,4 +15,10 @@ abstract class ApiService {
 
   @GET('/api/v1/me')
   Future<MeResponse> getMe();
+
+  @GET('/api/v1/surveys?page[number]={pageNumber}&page[size]={pageSize}')
+  Future<SurveyListResponse> surveyList(
```
Comment (manh-t) [inline lib/api/api_service.dart]: @nmint8m please add the prefix verb to the method 🙋

File: lib/api/repository/survey_repository.dart [hunk 2]
```diff
@@ -0,0 +1,35 @@
+import 'package:injectable/injectable.dart';
+import 'package:kayla_flutter_ic/api/api_service.dart';
+import 'package:kayla_flutter_ic/api/exception/network_exceptions.dart';
+import 'package:kayla_flutter_ic/model/survey.dart';
+
+abstract class SurveyRepository {
+  Future<List<Survey>> surveyList(
```
Comment (manh-t) [inline lib/api/repository/survey_repository.dart]: same

File: lib/usecases/survey/survey_list_use_case.dart [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import 'dart:async';
+import 'package:injectable/injectable.dart';
+import 'package:kayla_flutter_ic/api/repository/survey_repository.dart';
+import 'package:kayla_flutter_ic/model/survey.dart';
+import 'package:kayla_flutter_ic/usecases/base/base_use_case.dart';
+import 'package:kayla_flutter_ic/usecases/survey/survey_list_params.dart';
+
+@Injectable()
+class SurveyListUseCase extends UseCase<List<Survey>, SurveyListParams> {
```
Comment (manh-t) [inline lib/usecases/survey/survey_list_use_case.dart]: `GetSurveyListUseCase

File: lib/views/home/home_view_model.dart [hunk 4]
```diff
@@ -17,8 +20,12 @@ class HomeViewModel extends StateNotifier<HomeState> {
   final StreamController<List<String>> _surveyListStream = StreamController();
 
   final ProfileUseCase _profileUseCase;
```
Comment (manh-t) [inline lib/views/home/home_view_model.dart]: It seems other use cases are missing the prefix as well 😅 our use case always has the verb to understand the right behavior of the use case FYI
image

File: lib/api/api_service.dart [hunk 5]
```diff
@@ -14,4 +15,10 @@ abstract class ApiService {
 
   @GET('/api/v1/me')
   Future<MeResponse> getMyProfile();
+
+  @GET('/api/v1/surveys?page[number]={pageNumber}&page[size]={pageSize}')
+  Future<SurveyListResponse> getSurveyList(
```
Comment (doannimble) [inline lib/api/api_service.dart]: Optional: we can rename it to `getSurveys`

File: lib/api/response/survey_list_response.dart [hunk 6]
```diff
@@ -0,0 +1,18 @@
+import 'package:json_annotation/json_annotation.dart';
+import 'package:kayla_flutter_ic/api/response/base_response_converter.dart';
+import 'package:kayla_flutter_ic/api/response/survey_response.dart';
+
+part 'survey_list_response.g.dart';
+
+@JsonSerializable()
+class SurveyListResponse {
```
Comment (doannimble) [inline lib/api/response/survey_list_response.dart]:
```suggestion
class SurveysResponse {
```

**Review by doannimble — COMMENTED** [review]
rest lgtm

**Review by manh-t — APPROVED** [review]

File: lib/usecases/survey/get_survey_list_use_case.dart [hunk 7]
```diff
@@ -0,0 +1,28 @@
+import 'dart:async';
+import 'package:injectable/injectable.dart';
+import 'package:kayla_flutter_ic/api/repository/survey_repository.dart';
+import 'package:kayla_flutter_ic/api/response/surveys_response.dart';
+import 'package:kayla_flutter_ic/usecases/base/base_use_case.dart';
+import 'package:kayla_flutter_ic/usecases/survey/survey_list_params.dart';
+
+@Injectable()
+class GetSurveyListUseCase extends UseCase<SurveysResponse, SurveyListParams> {
```
Comment (doannimble) [inline lib/usecases/survey/get_survey_list_use_case.dart]:
```suggestion
class GetSurveysUseCase extends UseCase<SurveysResponse, SurveyListParams> {
```

File: lib/usecases/survey/survey_list_params.dart [hunk 8]
```diff
@@ -0,0 +1,9 @@
+class SurveyListParams {
```
Comment (doannimble) [inline lib/usecases/survey/survey_list_params.dart]:
```suggestion
class SurveysParams {
```

File: lib/usecases/survey/survey_list_use_case.dart [hunk 9]
```diff
@@ -0,0 +1,28 @@
+import 'dart:async';
+import 'package:injectable/injectable.dart';
+import 'package:kayla_flutter_ic/api/repository/survey_repository.dart';
+import 'package:kayla_flutter_ic/api/response/surveys_response.dart';
+import 'package:kayla_flutter_ic/usecases/base/base_use_case.dart';
+import 'package:kayla_flutter_ic/usecases/survey/survey_list_params.dart';
+
+@Injectable()
+class SurveyListUseCase extends UseCase<SurveysResponse, SurveyListParams> {
```
Comment (doannimble) [inline lib/usecases/survey/survey_list_use_case.dart]:
```suggestion
class SurveysUseCase extends UseCase<SurveysResponse, SurveyListParams> {
```

**Pull request merged**